File: src/voltis/routes/files.py
import io
import zipfile
from collections.abc import AsyncIterator
from pathlib import Path

import anyio
import anyio.from_thread
import anyio.to_thread
from anyio.streams.memory import MemoryObjectSendStream
from fastapi import APIRouter, HTTPException
from fastapi.responses import FileResponse, Response
from fastapi.responses import StreamingResponse as StreamingResponse
from pydantic import BaseModel
from sqlalchemy import func, select
from sqlalchemy.orm import load_only

from voltis.components.epub import list_chapters, read_chapter
from voltis.db.models import Content
from voltis.routes._providers import RbProvider, UserProvider
from voltis.utils.cover_cache import read_content_cover, read_content_file
from voltis.utils.misc import notnone

router = APIRouter()


@router.get("/cover/{content_id}")
async def get_cover(
    rb: RbProvider,
    _user: UserProvider,
    content_id: str,
    v: str | None = None,
) -> Response:
    async with rb.get_asession() as session:
        content = await session.get(Content, content_id)
        if not content:
            raise HTTPException(status_code=404, detail="Content not found")

        if not content.cover_uri:
            raise HTTPException(status_code=404, detail="Content has no cover")

        data, media_type = await anyio.to_thread.run_sync(read_content_cover, content)

        headers = {}
        if v:
            headers["Cache-Control"] = "public, max-age=31536000, immutable"
        return Response(content=data, media_type=media_type, headers=headers)


@router.get("/comic-page/{content_id}/{page_index}")
async def get_page(
    rb: RbProvider,
    _user: UserProvider,
    content_id: str,
    page_index: int,
    v: str | None = None,
) -> Response:
    async with rb.get_asession() as session:
        content = await session.get(Content, content_id)
        if not content:
            raise HTTPException(status_code=404, detail="Content not found")

        file_data = content.file_data
        if not content.file_uri or not file_data or "pages" not in file_data:
            raise HTTPException(status_code=404, detail="Content has no pages")

        if page_index < 0 or page_index >= len(file_data["pages"]):
            raise HTTPException(status_code=404, detail="Page index out of range")

        page_name = file_data["pages"][page_index]
        file_path = Path(content.file_uri)
        page_uri = file_path / page_name[0]

        data, media_type = await anyio.to_thread.run_sync(read_content_file, page_uri.as_posix())

        headers = {}
        if v:
            headers["Cache-Control"] = "public, max-age=31536000, immutable"
        return Response(content=data, media_type=media_type, headers=headers)


class ChapterResponse(BaseModel):
    id: str
    href: str
    title: str | None
    linear: bool


@router.get("/book-chapters/{content_id}")
async def get_book_chapters(
    rb: RbProvider,
    _user: UserProvider,
    content_id: str,
) -> list[ChapterResponse]:
    async with rb.get_asession() as session:
        content = await session.get(Content, content_id)
        if not content or not content.file_uri:
            raise HTTPException(status_code=404, detail="Content not found")

        file_path = Path(content.file_uri)
        if not file_path.suffix.lower() == ".epub":
            raise HTTPException(status_code=400, detail="Content is not an EPUB")

        try:
            chapters = await anyio.to_thread.run_sync(list_chapters, file_path)
        except (ValueError, FileNotFoundError) as e:
            raise HTTPException(status_code=500, detail=str(e))

        return [
            ChapterResponse(id=ch.id, href=ch.href, title=ch.title, linear=ch.linear)
            for ch in chapters
        ]


@router.get("/book-chapter/{content_id}")
async def get_book_chapter(
    rb: RbProvider,
    _user: UserProvider,
    content_id: str,
    href: str,
) -> Response:
    async with rb.get_asession() as session:
        content = await session.get(Content, content_id)
        if not content or not content.file_uri:
            raise HTTPException(status_code=404, detail="Content not found")

        file_path = Path(content.file_uri)
        if not file_path.suffix.lower() == ".epub":
            raise HTTPException(status_code=400, detail="Content is not an EPUB")

        try:
            chapter_content = await anyio.to_thread.run_sync(read_chapter, file_path, href)
        except FileNotFoundError:
            raise HTTPException(status_code=404, detail="Chapter not found")

        return Response(content=chapter_content, media_type="application/xhtml+xml")


@router.get("/book-resource/{content_id}")
async def get_book_resource(
    rb: RbProvider,
    _user: UserProvider,
    content_id: str,
    path: str,
) -> Response:
    """Serve a resource (image, CSS, etc.) from inside an EPUB file."""
    async with rb.get_asession() as session:
        content = await session.get(Content, content_id)
        if not content or not content.file_uri:
            raise HTTPException(status_code=404, detail="Content not found")

        file_path = Path(content.file_uri).resolve()
        if not file_path.suffix.lower() == ".epub":
            raise HTTPException(status_code=400, detail="Content is not an EPUB")

        final_path = (file_path / path).resolve().as_posix()
        if not final_path.startswith(file_path.as_posix() + "/"):
            raise HTTPException(status_code=400, detail="Invalid resource path")

        data, media_type = await anyio.to_thread.run_sync(read_content_file, final_path)
        return Response(content=data, media_type=media_type)


class DownloadInfoResponse(BaseModel):
    file_count: int
    total_size: int | None


@router.get("/download-info/{content_id}")
async def get_download_info(
    rb: RbProvider,
    _user: UserProvider,
    content_id: str,
) -> DownloadInfoResponse:
    async with rb.get_asession() as session:
        content = await session.get(Content, content_id)
        if not content:
            raise HTTPException(status_code=404, detail="Content not found")

        if content.type in ("comic", "book"):
            if not content.file_uri:
                raise HTTPException(status_code=404, detail="Content has no file")
            return DownloadInfoResponse(
                file_count=1,
                total_size=content.file_size,
            )

        # Series: aggregate children
        result = await session.execute(
            select(
                func.count(Content.id),
                func.sum(Content.file_size),
            ).where((Content.parent_id == content_id) & (Content.file_uri.isnot(None)))
        )
        file_count, total_size = result.one()
        if not file_count:
            raise HTTPException(status_code=404, detail="No downloadable files")

        return DownloadInfoResponse(
            file_count=file_count,
            total_size=total_size,
        )


@router.get("/download/{content_id}")
async def download(
    rb: RbProvider,
    _user: UserProvider,
    content_id: str,
) -> Response:
    async with rb.get_asession() as session:
        content = await session.get(Content, content_id)
        if not content:
            raise HTTPException(status_code=404, detail="Content not found")

        if content.type in ("comic", "book"):
            if not content.file_uri:
                raise HTTPException(status_code=404, detail="Content has no file")
            file_path = Path(content.file_uri)
            return FileResponse(
                file_path,
                filename=file_path.name,
                media_type="application/octet-stream",
            )

        # Series: stream a ZIP of all children's files
        result = await session.execute(
            select(Content)
            .options(load_only(Content.file_uri, Content.uri_part))
            .where((Content.parent_id == content_id) & (Content.file_uri.isnot(None)))
            .order_by(Content.order)
        )
        children = result.scalars().all()
        if not children:
            raise HTTPException(status_code=404, detail="No downloadable files")

        file_paths = [(Path(notnone(c.file_uri)), Path(notnone(c.file_uri)).name) for c in children]
        zip_filename = f"{content.uri_part}.zip"

        async def stream_zip() -> AsyncIterator[bytes]:
            send_stream, receive_stream = anyio.create_memory_object_stream[bytes](8)

            async def build_zip() -> None:
                async with send_stream:
                    await anyio.to_thread.run_sync(_write_zip_to_stream, file_paths, send_stream)

            async with anyio.create_task_group() as tg:
                tg.start_soon(build_zip)
                async with receive_stream:
                    async for chunk in receive_stream:
                        yield chunk

        headers = {
            "Content-Disposition": f'attachment; filename="{zip_filename}"',
        }
        return StreamingResponse(
            stream_zip(),
            media_type="application/zip",
            headers=headers,
        )


def _write_zip_to_stream(
    file_paths: list[tuple[Path, str]],
    send_stream: MemoryObjectSendStream[bytes],
) -> None:
    """Build a ZIP in a worker thread, sending chunks via the stream."""
    buf = io.BytesIO()
    sent = 0
    with zipfile.ZipFile(buf, "w", compression=zipfile.ZIP_STORED) as zf:
        for file_path, arcname in file_paths:
            zf.write(file_path, arcname)
            pos = buf.tell()
            if pos > sent:
                buf.seek(sent)
                anyio.from_thread.run(send_stream.send, buf.read(pos - sent))
                sent = pos
    # Final flush for central directory
    buf.seek(sent)
    remainder = buf.read()
    if remainder:
        anyio.from_thread.run(send_stream.send, remainder)

File: backend/routes/files.go
package routes

import (
	"archive/zip"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"mime"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"

	"voltis/config"
	"voltis/db"
	"voltis/lib/archive"
	"voltis/lib/epub"
	"voltis/models"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/labstack/echo/v4"
)

type FileRoutes struct {
	pool *pgxpool.Pool
}

func (fr *FileRoutes) Register(g *echo.Group) {
	g.GET("/cover/:content_id", fr.getCover)
	g.GET("/comic-page/:content_id/:page_index", fr.getComicPage)
	g.GET("/book-chapters/:content_id", fr.getBookChapters)
	g.GET("/book-chapter/:content_id", fr.getBookChapter)
	g.GET("/book-resource/:content_id", fr.getBookResource)
	g.GET("/download-info/:content_id", fr.getDownloadInfo)
	g.GET("/download/:content_id", fr.download)
}

func (fr *FileRoutes) getCover(c echo.Context) error {
	if _, err := requireUser(c); err != nil {
		return err
	}

	ctx := reqCtx(c)
	contentID := c.Param("content_id")

	content, err := db.SelectOne[models.Content](ctx, fr.pool, "SELECT * FROM content WHERE id = $1", contentID)
	if errors.Is(err, pgx.ErrNoRows) {
		return echo.NewHTTPError(http.StatusNotFound, "Content not found")
	}
	if err != nil {
		return err
	}

	if content.CoverURI == nil {
		return echo.NewHTTPError(http.StatusNotFound, "Content has no cover")
	}

	data, mediaType, err := readCover(content)
	if err != nil {
		return err
	}

	if c.QueryParam("v") != "" {
		c.Response().Header().Set("Cache-Control", "public, max-age=31536000, immutable")
	}
	return c.Blob(http.StatusOK, mediaType, data)
}

func (fr *FileRoutes) getComicPage(c echo.Context) error {
	if _, err := requireUser(c); err != nil {
		return err
	}

	ctx := reqCtx(c)
	contentID := c.Param("content_id")

	var pageIndex int
	if _, err := fmt.Sscanf(c.Param("page_index"), "%d", &pageIndex); err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "invalid page index")
	}

	content, err := db.SelectOne[models.Content](ctx, fr.pool, "SELECT * FROM content WHERE id = $1", contentID)
	if errors.Is(err, pgx.ErrNoRows) {
		return echo.NewHTTPError(http.StatusNotFound, "Content not found")
	}
	if err != nil {
		return err
	}

	if content.FileURI == nil || content.FileData == nil {
		return echo.NewHTTPError(http.StatusNotFound, "Content has no pages")
	}

	var fileData struct {
		Pages []json.RawMessage `json:"pages"`
	}
	if err := json.Unmarshal(content.FileData, &fileData); err != nil || len(fileData.Pages) == 0 {
		return echo.NewHTTPError(http.StatusNotFound, "Content has no pages")
	}

	if pageIndex < 0 || pageIndex >= len(fileData.Pages) {
		return echo.NewHTTPError(http.StatusNotFound, "Page index out of range")
	}

	// Pages are stored as [name, width, height] tuples
	var pageTuple []json.RawMessage
	if err := json.Unmarshal(fileData.Pages[pageIndex], &pageTuple); err != nil || len(pageTuple) == 0 {
		return echo.NewHTTPError(http.StatusNotFound, "Invalid page data")
	}
	var pageName string
	if err := json.Unmarshal(pageTuple[0], &pageName); err != nil {
		return echo.NewHTTPError(http.StatusNotFound, "Invalid page data")
	}

	pageURI := filepath.Join(*content.FileURI, pageName)
	data, mediaType, err := readContentFile(pageURI)
	if err != nil {
		return err
	}

	if c.QueryParam("v") != "" {
		c.Response().Header().Set("Cache-Control", "public, max-age=31536000, immutable")
	}
	return c.Blob(http.StatusOK, mediaType, data)
}

type chapterResponse struct {
	ID     string  `json:"id"`
	Href   string  `json:"href"`
	Title  *string `json:"title"`
	Linear bool    `json:"linear"`
}

func (fr *FileRoutes) getBookChapters(c echo.Context) error {
	if _, err := requireUser(c); err != nil {
		return err
	}

	ctx := reqCtx(c)
	contentID := c.Param("content_id")

	content, err := db.SelectOne[models.Content](ctx, fr.pool, "SELECT * FROM content WHERE id = $1", contentID)
	if errors.Is(err, pgx.ErrNoRows) {
		return echo.NewHTTPError(http.StatusNotFound, "Content not found")
	}
	if err != nil {
		return err
	}

	if content.FileURI == nil {
		return echo.NewHTTPError(http.StatusNotFound, "Content not found")
	}
	if !strings.HasSuffix(strings.ToLower(*content.FileURI), ".epub") {
		return echo.NewHTTPError(http.StatusBadRequest, "Content is not an EPUB")
	}

	chapters, err := epub.ListChapters(*content.FileURI)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	result := make([]chapterResponse, len(chapters))
	for i, ch := range chapters {
		result[i] = chapterResponse{
			ID:     ch.ID,
			Href:   ch.Href,
			Linear: ch.Linear,
		}
		if ch.Title != "" {
			result[i].Title = &ch.Title
		}
	}
	return c.JSON(http.StatusOK, result)
}

func (fr *FileRoutes) getBookChapter(c echo.Context) error {
	if _, err := requireUser(c); err != nil {
		return err
	}

	ctx := reqCtx(c)
	contentID := c.Param("content_id")
	href := c.QueryParam("href")

	content, err := db.SelectOne[models.Content](ctx, fr.pool, "SELECT * FROM content WHERE id = $1", contentID)
	if errors.Is(err, pgx.ErrNoRows) {
		return echo.NewHTTPError(http.StatusNotFound, "Content not found")
	}
	if err != nil {
		return err
	}

	if content.FileURI == nil {
		return echo.NewHTTPError(http.StatusNotFound, "Content not found")
	}
	if !strings.HasSuffix(strings.ToLower(*content.FileURI), ".epub") {
		return echo.NewHTTPError(http.StatusBadRequest, "Content is not an EPUB")
	}

	chapterContent, err := epub.ReadChapter(*content.FileURI, href)
	if err != nil {
		return echo.NewHTTPError(http.StatusNotFound, "Chapter not found")
	}

	return c.Blob(http.StatusOK, "application/xhtml+xml", []byte(chapterContent))
}

func (fr *FileRoutes) getBookResource(c echo.Context) error {
	if _, err := requireUser(c); err != nil {
		return err
	}

	ctx := reqCtx(c)
	contentID := c.Param("content_id")
	resourcePath := c.QueryParam("path")

	content, err := db.SelectOne[models.Content](ctx, fr.pool, "SELECT * FROM content WHERE id = $1", contentID)
	if errors.Is(err, pgx.ErrNoRows) {
		return echo.NewHTTPError(http.StatusNotFound, "Content not found")
	}
	if err != nil {
		return err
	}

	if content.FileURI == nil {
		return echo.NewHTTPError(http.StatusNotFound, "Content not found")
	}

	if !strings.HasSuffix(strings.ToLower(*content.FileURI), ".epub") {
		return echo.NewHTTPError(http.StatusBadRequest, "Content is not an EPUB")
	}

	// Path traversal protection
	fileBase, _ := filepath.Abs(*content.FileURI)
	resolved, _ := filepath.Abs(filepath.Join(*content.FileURI, resourcePath))
	if !strings.HasPrefix(resolved, fileBase+string(filepath.Separator)) {
		return echo.NewHTTPError(http.StatusBadRequest, "Invalid resource path")
	}

	data, mediaType, err := readContentFile(resolved)
	if err != nil {
		return err
	}
	return c.Blob(http.StatusOK, mediaType, data)
}

type downloadInfoResponse struct {
	FileCount int  `json:"file_count"`
	TotalSize *int `json:"total_size"`
}

func (fr *FileRoutes) getDownloadInfo(c echo.Context) error {
	if _, err := requireUser(c); err != nil {
		return err
	}

	ctx := reqCtx(c)
	contentID := c.Param("content_id")

	content, err := db.SelectOne[models.Content](ctx, fr.pool, "SELECT * FROM content WHERE id = $1", contentID)
	if errors.Is(err, pgx.ErrNoRows) {
		return echo.NewHTTPError(http.StatusNotFound, "Content not found")
	}
	if err != nil {
		return err
	}

	if content.Type == "comic" || content.Type == "book" {
		if content.FileURI == nil {
			return echo.NewHTTPError(http.StatusNotFound, "Content has no file")
		}
		return c.JSON(http.StatusOK, downloadInfoResponse{
			FileCount: 1,
			TotalSize: content.FileSize,
		})
	}

	// Series: aggregate children
	var stats struct {
		FileCount int  `db:"file_count"`
		TotalSize *int `db:"total_size"`
	}
	err = fr.pool.QueryRow(ctx, `
		SELECT COUNT(*) AS file_count, SUM(file_size)::int AS total_size
		FROM content WHERE parent_id = $1 AND file_uri IS NOT NULL
	`, contentID).Scan(&stats.FileCount, &stats.TotalSize)
	if err != nil {
		return err
	}
	if stats.FileCount == 0 {
		return echo.NewHTTPError(http.StatusNotFound, "No downloadable files")
	}

	return c.JSON(http.StatusOK, downloadInfoResponse{
		FileCount: stats.FileCount,
		TotalSize: stats.TotalSize,
	})
}

func (fr *FileRoutes) download(c echo.Context) error {
	if _, err := requireUser(c); err != nil {
		return err
	}

	ctx := reqCtx(c)
	contentID := c.Param("content_id")

	content, err := db.SelectOne[models.Content](ctx, fr.pool, "SELECT * FROM content WHERE id = $1", contentID)
	if errors.Is(err, pgx.ErrNoRows) {
		return echo.NewHTTPError(http.StatusNotFound, "Content not found")
	}
	if err != nil {
		return err
	}

	if content.Type == "comic" || content.Type == "book" {
		if content.FileURI == nil {
			return echo.NewHTTPError(http.StatusNotFound, "Content has no file")
		}
		return c.File(*content.FileURI)
	}

	// Series: stream a ZIP of all children's files
	type childFile struct {
		FileURI string `db:"file_uri"`
		URIPart string `db:"uri_part"`
	}
	children, err := db.Select[childFile](ctx, fr.pool, `
		SELECT file_uri, uri_part FROM content
		WHERE parent_id = $1 AND file_uri IS NOT NULL
		ORDER BY "order"
	`, contentID)
	if err != nil {
		return err
	}
	if len(children) == 0 {
		return echo.NewHTTPError(http.StatusNotFound, "No downloadable files")
	}

	zipFilename := content.URIPart + ".zip"
	c.Response().Header().Set("Content-Type", "application/zip")
	c.Response().Header().Set("Content-Disposition", fmt.Sprintf(`attachment; filename="%s"`, zipFilename))
	c.Response().WriteHeader(http.StatusOK)

	zw := zip.NewWriter(c.Response())
	defer func() { _ = zw.Close() }()

	for _, child := range children {
		name := filepath.Base(child.FileURI)
		w, err := zw.CreateHeader(&zip.FileHeader{
			Name:   name,
			Method: zip.Store,
		})
		if err != nil {
			return err
		}

		f, err := os.Open(child.FileURI)
		if err != nil {
			return err
		}
		_, err = io.Copy(w, f)
		_ = f.Close()
		if err != nil {
			return err
		}
	}

	return nil
}

// File reading helpers

var archiveExtensions = map[string]bool{
	".zip": true, ".cbz": true, ".cbr": true, ".rar": true, ".epub": true, ".pdf": true,
}

var pdfPagePattern = regexp.MustCompile(`^p(\d+)$`)

func readContentFile(uri string) ([]byte, string, error) {
	// Try as a regular file first
	if data, err := os.ReadFile(uri); err == nil {
		mediaType := guessMediaType(uri)
		return data, mediaType, nil
	}

	// Walk up the path to find an archive
	archivePath, innerPath := findArchiveAndInnerPath(uri)
	if archivePath == "" {
		return nil, "", echo.NewHTTPError(http.StatusNotFound, "File not found")
	}

	// PDF page rendering via pdftoppm
	if strings.ToLower(filepath.Ext(archivePath)) == ".pdf" {
		return readPDFPage(archivePath, innerPath)
	}

	a, err := archive.Open(archivePath)
	if err != nil {
		return nil, "", echo.NewHTTPError(http.StatusNotFound, "File not found")
	}
	defer func() { _ = a.Close() }()

	data, err := a.ReadFile(innerPath)
	if err != nil {
		return nil, "", echo.NewHTTPError(http.StatusNotFound, "File not found")
	}

	mediaType := guessMediaType(innerPath)
	return data, mediaType, nil
}

func readPDFPage(pdfPath, innerPath string) ([]byte, string, error) {
	m := pdfPagePattern.FindStringSubmatch(innerPath)
	if m == nil {
		return nil, "", echo.NewHTTPError(http.StatusBadRequest, "Invalid PDF page identifier")
	}
	page := m[1]

	cmd := exec.Command("pdftoppm",
		"-r", "250",
		"-jpeg", "-jpegopt", "quality=90",
		"-singlefile",
		"-f", page, "-l", page,
		pdfPath,
	)
	data, err := cmd.Output()
	if err != nil {
		return nil, "", echo.NewHTTPError(http.StatusInternalServerError, "PDF rendering failed")
	}
	return data, "image/jpeg", nil
}

func readCover(content models.Content) ([]byte, string, error) {
	if content.CoverURI == nil {
		return nil, "", echo.NewHTTPError(http.StatusNotFound, "Content has no cover")
	}

	cfg := config.Get()
	cacheDir := filepath.Join(cfg.CacheDir, "covers")
	cachePath := filepath.Join(cacheDir, content.ID+".jpg")

	// Check cache
	if info, err := os.Stat(cachePath); err == nil {
		if content.FileMtime == nil || !info.ModTime().Before(*content.FileMtime) {
			data, err := os.ReadFile(cachePath)
			if err == nil {
				return data, "image/jpeg", nil
			}
		}
	}

	// Read from source
	data, _, err := readContentFile(*content.CoverURI)
	if err != nil {
		return nil, "", err
	}

	// Cache (best-effort)
	_ = os.MkdirAll(cacheDir, 0o755)
	_ = os.WriteFile(cachePath, data, 0o644)

	return data, guessMediaType(*content.CoverURI), nil
}

func findArchiveAndInnerPath(uri string) (string, string) {
	parts := strings.Split(filepath.ToSlash(uri), "/")
	for i := len(parts) - 1; i > 0; i-- {
		candidate := strings.Join(parts[:i], "/")
		ext := strings.ToLower(filepath.Ext(candidate))
		if !archiveExtensions[ext] {
			continue
		}
		if info, err := os.Stat(candidate); err == nil && !info.IsDir() {
			innerPath := strings.Join(parts[i:], "/")
			return candidate, innerPath
		}
	}
	return "", ""
}

func guessMediaType(name string) string {
	ext := filepath.Ext(name)
	if ext == "" {
		return "application/octet-stream"
	}
	mt := mime.TypeByExtension(ext)
	if mt == "" {
		return "application/octet-stream"
	}
	return mt
}
